src/processes.jl

```diff
@@ -119,8 +119,15 @@ end
 # performance for `map!` to be comparable with the native loop.
 # See: ../benchmark/bench_filter_map_map!.jl
 
-Base.mapfoldl(xform::Transducer, f, init, itr) =
-    unreduced(transduce(xform, f, init, itr))
+function mapfoldl_init(xform, step, itr)
+    T = outtype(xform, ieltype(itr))
```

Comment:
This actually does an `outtype`-inference twice.  I think a better place to invoke `Base.reduce_empty(step, T)` is after constructing a `Reduction` (i.e., transducers + `step`) which requires evoking `outtype` anyway.  One idea is to define a singleton `Unspecified` and pass it around as `init`:

```julia
struct Unspecified end

Base.mapfoldl(xform::Transducer, f, itr; init = Unspecified()) =
    unreduced(transduce(xform, f, init, itr))

@inline function transduce(rf::Reduction, init, coll)
    if init isa Unspecified
        # get `step` and `T` from `rf`, then...
        val = Base.reduce_empty(step, T)
    else
        val = init
    end
    return __foldl__(rf, start(rf, val), coll)
end
```

It's an optimization that can be done later.  But, if you feel like doing it, please go ahead and do it.